feat: update auth provider type validation

theref opened the pull request:
Updates AuthProviderType to handle multiple provider types for USER_ADDRESS_PARAM_DEFAULT while maintaining single type for external EIP4361. Updates related tests to properly map auth provider keys.

theref committed:
feat: update auth provider type validation

Updates AuthProviderType to handle multiple provider types for USER_ADDRESS_PARAM_DEFAULT while
maintaining single type for external EIP4361. Updates related tests to properly map auth provider
keys.

derekpierre commented on packages/taco/test/conditions/context.test.ts:
```diff
@@ -501,17 +475,17 @@ describe('context', () => {
-          USER_ADDRESS_PARAM_EIP1271,
-          authProviders[USER_ADDRESS_PARAM_EIP1271],
+          USER_ADDRESS_PARAM_DEFAULT,
+          authProviders["EIP4361"],
```
This original goal was for variety of the tests., so that we aren't just using EIP4361 all the time.

```suggestion
          authProviders["EIP1271"],
```

derekpierre commented on packages/taco/src/conditions/context/context.ts:
```diff
@@ -40,19 +39,17 @@ const ERR_AUTH_PROVIDER_NOT_NEEDED_FOR_CONTEXT_PARAM = (param: string) =>
-  | typeof EIP1271AuthProvider;
+  | (typeof EIP4361AuthProvider | typeof EIP1271AuthProvider)[]
+  | typeof SingleSignOnEIP4361AuthProvider;
+
 
 const EXPECTED_AUTH_PROVIDER_TYPES: Record<string, AuthProviderType> = {
-  [USER_ADDRESS_PARAM_DEFAULT]: EIP4361AuthProvider,
+  [USER_ADDRESS_PARAM_DEFAULT]: [EIP4361AuthProvider, EIP1271AuthProvider],
   [USER_ADDRESS_PARAM_EXTERNAL_EIP4361]: SingleSignOnEIP4361AuthProvider,
-  [USER_ADDRESS_PARAM_EIP1271]: EIP1271AuthProvider,
 };
```
I think this got more complicated than needed:

```typescript
type AuthProviderType =
  | typeof EIP4361AuthProvider
  | typeof EIP1271AuthProvider
  | typeof SingleSignOnEIP4361AuthProvider;

const EXPECTED_AUTH_PROVIDER_TYPES: Record<string, AuthProviderType[]> = {
  [USER_ADDRESS_PARAM_DEFAULT]: [EIP4361AuthProvider, EIP1271AuthProvider],
  [USER_ADDRESS_PARAM_EXTERNAL_EIP4361]: [SingleSignOnEIP4361AuthProvider],
};
```

derekpierre commented on packages/taco/src/conditions/context/context.ts:
```diff
@@ -222,16 +219,19 @@ export class ConditionContext {
       );
     }
+    const expectedType = EXPECTED_AUTH_PROVIDER_TYPES[contextParam];
+    const isValid = Array.isArray(expectedType)
+      ? expectedType.some(type => authProvider instanceof type)
+      : authProvider instanceof expectedType;
 
-    if (!(authProvider instanceof EXPECTED_AUTH_PROVIDER_TYPES[contextParam])) {
+    if (!isValid) {
```
Given the above comment about the types, then we can do the following:
```typescript
  public addAuthProvider(contextParam: string, authProvider: AuthProvider) {
    if (!(contextParam in EXPECTED_AUTH_PROVIDER_TYPES)) {
      throw new Error(
        ERR_AUTH_PROVIDER_NOT_NEEDED_FOR_CONTEXT_PARAM(contextParam),
      );
    }
    const expectedTypes = EXPECTED_AUTH_PROVIDER_TYPES[contextParam];
    if (!expectedTypes.some((type) => authProvider instanceof type)) {
      throw new Error(
        ERR_INVALID_AUTH_PROVIDER_TYPE(contextParam, typeof authProvider),
      );
    }

    this.authProviders[contextParam] = authProvider;
  }
```

theref replied:
yeah it definitely simplifies things if we just always make things arrays, even if they only contain one item

derekpierre commented on packages/taco/test/conditions/context.test.ts:
```diff
@@ -212,36 +204,17 @@ describe('context', () => {
-      USER_ADDRESS_PARAM_EIP1271,
-    ])('it supports just one provider at a time', async (userAddressParam) => {
+      [USER_ADDRESS_PARAM_DEFAULT, "EIP4361"],
+      [USER_ADDRESS_PARAM_EXTERNAL_EIP4361, "SSO4361"],
+    ])('it supports just one provider at a time', async (userAddressParam, authProviderKey) => {
```
Missing test case here:

```suggestion
      [USER_ADDRESS_PARAM_DEFAULT, "EIP1271"],
      [USER_ADDRESS_PARAM_EXTERNAL_EIP4361, "SSO4361"],
    ])('it supports just one provider at a time', async (userAddressParam, scheme) => {
```

derekpierre commented on packages/taco/test/conditions/context.test.ts:
```diff
@@ -253,7 +226,7 @@ describe('context', () => {
       conditionContext.addAuthProvider(
         userAddressParam,
-        authProviders[userAddressParam],
+        authProviders[authProviderKey],
```
```suggestion
        authProviders[scheme],
```

derekpierre commented on packages/taco/test/conditions/context.test.ts:
```diff
@@ -303,7 +276,7 @@ describe('context', () => {
     }
 
-    async function makeAuthSignature(userAddressParam: string) {
+    async function makeAuthSignature(userAddressParam: string, scheme: string) {
```
Does it make sense to simplify this?

```suggestion
    async function makeAuthSignature(scheme: string) {
      const userAddressParam = scheme === "SSO4361" ? USER_ADDRESS_PARAM_EXTERNAL_EIP4361 : USER_ADDRESS_PARAM_DEFAULT

```

theref replied:
i'm not sure on this suggestion - it feels like it could get messy when we add new auth types

derekpierre replied:
Maybe you could expand on "messy" a bit more to help me understand?

I may be missing something, but, at least currently there is one-to-one relationship between scheme and userAddressParam i.e. if we know the scheme we know the relevant user address param for the signature. I question using `userAddressParam` as a parameter because having to provide both is redundant. So the question may be, is there a world where we continue to pass only a scheme, but don't know the userAddressParam to use? 

If you are thinking about other types of `AuthSignature` (non-user address param), then perhaps the helper method should be changed to `makeUserAddressAuthSignature(scheme: string)`.

theref replied:
i think this relates to your comment below - if the scheme is `SSO4361`, the `userAddressParam` could be `USER_ADDRESS_PARAM_EXTERNAL_EIP4361` or `USER_ADDRESS_PARAM_DEFAULT`

I'd just prefer not to maintain a mapping, and instead allow different combos to be passed in

derekpierre replied:
Actually, based on my subsequent suggestion here - https://github.com/derekpierre/taco-web/pull/1#discussion_r1950892386, it now makes sense to keep userAddressParam. Ignore the above.

derekpierre replied:
> i think this relates to your comment below - if the scheme is `SSO4361`, the `userAddressParam` could be `USER_ADDRESS_PARAM_EXTERNAL_EIP4361` or `USER_ADDRESS_PARAM_DEFAULT`
> 
> I'd just prefer not to maintain a mapping, and instead allow different combos to be passed in

Haha jinx. Indeed - totally right.

theref committed:
Apply PR suggestion to simplify AuthProviderType

derekpierre commented on packages/taco/src/conditions/context/context.ts:
```diff
@@ -41,18 +40,16 @@ const ERR_AUTH_PROVIDER_NOT_NEEDED_FOR_CONTEXT_PARAM = (param: string) =>
-  [USER_ADDRESS_PARAM_EXTERNAL_EIP4361]: SingleSignOnEIP4361AuthProvider,
-  [USER_ADDRESS_PARAM_EIP1271]: EIP1271AuthProvider,
+const EXPECTED_AUTH_PROVIDER_TYPES: Record<string, AuthProviderType[]> = {
+  [USER_ADDRESS_PARAM_DEFAULT]: [EIP4361AuthProvider, EIP1271AuthProvider],
```
I missed this in my earlier suggestion, but I think we want to allow single-sign on for EIP4361 as well.

```suggestion
  [USER_ADDRESS_PARAM_DEFAULT]: [EIP4361AuthProvider, EIP1271AuthProvider, SingleSignOnEIP4361AuthProvider],
```

theref committed:
Add bogus provider for failure testing

Adds fakeBogusProvider export with proper ExternalProvider implementation to enable testing of authentication failure cases.

theref committed:
Fix fakeAuthProvider to implement AuthProvider properly
